lib/huginn_record_link_agent/record_link_agent.rb

```diff
@@ -393,53 +368,42 @@ def find_source_links(event, data)
       record = get_record_reference(target_system, target_type, target_id)
 
       if (record.nil?)
-        base_event = data['output_mode'].to_s == 'merge' ? event.payload.dup : {} if boolify(options['emit_events'])
 
-        payload = base_event.merge({
-          record_link: {
-            link_status: "500",
-            link_errors: [
-              {
-                message: "Unable to locate lookup record",
-                target_system: target_system,
-                target_type: target_type,
-                target_id: target_id,
-                source_system: source_system,
-                source_type: source_type
-              }
-            ]
+        payload = create_error_payload(
+          500,
+          {
+            message: "Unable to locate lookup record",
+            target_system: target_system,
+            target_type: target_type,
+            target_id: target_id,
+            source_system: source_system,
+            source_type: source_type
           }
-        })
+        )
 
-        create_event(payload: payload) if boolify(options['emit_events'])
-      else
-        source_records = record.source_records
+        emit(data, event, payload) if boolify(options['emit_events'])
 
-        source_records = source_records.where(ext_system: source_system) unless source_system.blank?
-        source_records = source_records.where(model_type: source_type) unless source_type.blank?
+      else
+        source_records = filter_links(record.source_records, source_system, source_type)
 
         if source_records.any?
           build_events(event, data, record, source_records) if boolify(options['emit_events'])
         else
-          base_event = data['output_mode'].to_s == 'merge' ? event.payload.dup : {} if boolify(options['emit_events'])
-
-          payload = base_event.merge({
-            record_link: {
-              link_status: "404",
-              link_errors: [
-                {
-                  message: "No match",
-                  target_system: target_system,
-                  target_type: target_type,
-                  target_id: target_id,
-                  source_system: source_system,
-                  source_type: source_type
-                }
-              ]
+
+          payload = create_error_payload(
```

Comment:
@mrogers-5s I'd rather have this method return the full structure, and be parameterized by the required data as individual arguments. That creates a more strictly defined interface.